### PR TITLE
Prevents players from getting antagonist if they can't qualify for any role beforehand.

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request
As the title says, you can no longer become an antagonist if you can't qualify for a role beforehand. How this works is that it runs job assignment before antagonist roles are handed out to see which players can qualify for a job role. It then undoes job assignments and then runs antagonist assignment.
This is to prevent players from setting a highly contested role to low priority so that they have a 0% chance of ever getting it, whilst still being eligible for every antag role.

It's not completely foolproof as it's not really possible to account for the fact that some players may become antagonists, thereby freeing up some job slots. Since it's an approximation, it means that players who put themselves at a high chance of returning back to lobby due to their preference choices have an overall lower chance of becoming an antagonist as they can be excluded from the antagonist roll due to the random nature of job selection.
If players want to prevent this from happening, simply set a lot of jobs to low/medium priority or adjust the "Return back to lobby" preference to any other preference.

## Why It's Good For The Game
Prevents players from rolling antag without ever intending to play the game as a non-antagonist.

## Changelog
:cl:
fix: Fixed players being able to roll antagonist without ever being eligible to play any role. Players who have their preferences set up so that they're likely to return to lobby when the round starts have a lowered chance of becoming antagonist.
/:cl:

### CODE REVIEW

**Comment by CCC23** [comment]
people do this?

**Comment by AtomTheProphet** [comment]
> people do this?

It's not the craziest thing that could happen.

**Comment by the-orange-cow** [comment]
Just so I understand this correctly, this only 'lowers' your antag chance if you have 'return to lobby' toggled AND you would have been returned to lobby that shift start already?

**Comment by DATA-xPUNGED** [comment]
> people do this?

I've seen someone getting banned for rolling antag by setting to return to lobby and low priority AI

**Comment by JupiterJaeden** [comment]
If you have only security roles set, can you still roll antag and be assigned a random non-sec job for it?

**Comment by Watermelon914** [comment]
> If you have only security roles set (or prisoner/captain), can you still roll antag and be assigned a random non-sec job for it?

yes

> Just so I understand this correctly, this only 'lowers' your antag chance if you have 'return to lobby' toggled AND you would have been returned to lobby that shift start already?

Basically, if you wouldn't have qualified for any role (aka it would send you back to the lobby), then it also means you wouldn't be able to roll antag.
What this means is that if you have your preferences set up in such a way that you get sent back to the lobby 50% of the time, it means that 50% of the time, you won't get antag no matter what, on top of your current chances of getting antag.
Statistically, this means you'll be getting antag less than other people.
Most people aren't like this though, they have a bunch of preferences set to medium or low. It'd primarily only be a problem for you if you have 1 job set to high that you may not get half the time and sent back to lobby as a result.

> people do this?

some people do whatever they can to maximize their antag chances without having to actually play the game non-antag

**Comment by necromanceranne** [comment]
> people do this?

It is a very obvious form of antag rolling that will definitely get you banned. It is very _blatant_, to say the least.

**Comment by xzero314** [comment]
Hey so I have a bit of an issue with this. I like to play Ai the most out of any job, I usually always 
set my job preference to ai or return to menu  or ai and sec or return to menu so that I can just pick something else if I dont get ai. This basically means I will never get to roll antag again due to the way I like to play and I dont think I am doing anything wrong? 

The way things work as is antag is rolled before job correct? So how is setting your job to one thing thats contested (ai) or menu increasing ones odds of rolling antag if the roll for ai is done after the antag roll?

**Comment by lessthnthree** [comment]
Is /mob/dead/new_player/proc/check_preferences() still required with this new check in place? It was doing similar.

**Comment by Watermelon914** [comment]
> Is /mob/dead/new_player/proc/check_preferences() still required with this new check in place? It was doing similar.

yes

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 1]
```diff
@@ -458,11 +458,17 @@ GLOBAL_LIST_EMPTY(dynamic_station_traits)
...
 	var/list/roundstart_rules = init_rulesets(/datum/dynamic_ruleset/roundstart)
 
+	SSjob.DivideOccupations(pure = TRUE, allow_all = TRUE)
 	for(var/i in GLOB.new_player_list)
 		var/mob/dead/new_player/player = i
 		if(player.ready == PLAYER_READY_TO_PLAY && player.mind && player.check_preferences())
-			roundstart_pop_ready++
-			candidates.Add(player)
+			if(is_unassigned_job(player.mind.assigned_role))
+				log_admin("[player.ckey] failed to qualify for any job and has [player.client.prefs.be_special.len] antag preferences enabled. They will be unable to qualify for any antagonist role.")
```
Comment (dragomagol) [inline code/game/gamemodes/dynamic/dynamic.dm:466]:
```suggestion
				log_admin("[player.ckey] failed to qualify for any job and has [player.client.prefs.be_special.len] antag preferences enabled. They will be unable to qualify for any roundstart antagonist role.")
```

I mentioned this briefly on discord, but I think that this wording should be explicit to say that they don't qualify for roundstart antags only.

**Comment by moocowswag** [comment]
Does this mean it wont force me to play shitty random jobs because i failed to roll scientist (thankyou) ive suicided countless times as antag because of this dogshit.

**Pull request merged**